tools-image/enki/cmd/build-iso.go

```diff
@@ -0,0 +1,126 @@
+			flags := cmd.Flags()
+
+			// Set this after parsing of the flags, so it fails on parsing and prints usage properly
+			cmd.SilenceUsage = true
+			cmd.SilenceErrors = true // Do not propagate errors down the line, we control them
+			spec, err := config.ReadBuildISO(cfg, flags)
+			if err != nil {
+				cfg.Logger.Errorf("invalid install command setup %v", err)
+				return err
+			}
+
+			if len(args) == 1 {
+				imgSource, err := v1.NewSrcFromURI(args[0])
+				if err != nil {
+					cfg.Logger.Errorf("not a valid rootfs source image argument: %s", args[0])
+					return err
+				}
+				spec.RootFS = []*v1.ImageSource{imgSource}
+			} else if len(spec.RootFS) == 0 {
+				errmsg := "rootfs source image for building ISO was not provided"
+				cfg.Logger.Errorf(errmsg)
+				return fmt.Errorf(errmsg)
+			}
+
+			// Repos and overlays can't be unmarshaled directly as they require
+			// to be merged on top and flags do not match any config value key
+			oRootfs, _ := flags.GetString("overlay-rootfs")
+			oUEFI, _ := flags.GetString("overlay-uefi")
+			oISO, _ := flags.GetString("overlay-iso")
+
+			if oRootfs != "" {
+				if ok, err := utils.Exists(cfg.Fs, oRootfs); ok {
+					spec.RootFS = append(spec.RootFS, v1.NewDirSrc(oRootfs))
+				} else {
+					cfg.Logger.Errorf("Invalid value for overlay-rootfs")
+					return fmt.Errorf("Invalid path '%s': %v", oRootfs, err)
+				}
+			}
+			if oUEFI != "" {
+				if ok, err := utils.Exists(cfg.Fs, oUEFI); ok {
+					spec.UEFI = append(spec.UEFI, v1.NewDirSrc(oUEFI))
+				} else {
+					cfg.Logger.Errorf("Invalid value for overlay-uefi")
+					return fmt.Errorf("Invalid path '%s': %v", oUEFI, err)
+				}
+			}
+			if oISO != "" {
+				if ok, err := utils.Exists(cfg.Fs, oISO); ok {
+					spec.Image = append(spec.Image, v1.NewDirSrc(oISO))
+				} else {
+					cfg.Logger.Errorf("Invalid value for overlay-iso")
+					return fmt.Errorf("Invalid path '%s': %v", oISO, err)
+				}
+			}
+
+			buildISO := action.NewBuildISOAction(cfg, spec)
+			err = buildISO.ISORun()
+			if err != nil {
+				cfg.Logger.Errorf(err.Error())
+				return err
+			}
+
+			return nil
+		},
+	}
+	root.AddCommand(c)
+	c.Flags().StringP("name", "n", "", "Basename of the generated ISO file")
+	c.Flags().StringP("output", "o", "", "Output directory (defaults to current directory)")
+	c.Flags().Bool("date", false, "Adds a date suffix into the generated ISO file")
+	c.Flags().String("overlay-rootfs", "", "Path of the overlayed rootfs data")
+	c.Flags().String("overlay-uefi", "", "Path of the overlayed uefi data")
+	c.Flags().String("overlay-iso", "", "Path of the overlayed iso data")
+	c.Flags().String("label", "", "Label of the ISO volume")
+	archType := newEnumFlag([]string{"x86_64", "arm64"}, "x86_64")
+	c.Flags().Bool("squash-no-compression", true, "Disable squashfs compression.")
+	c.Flags().VarP(archType, "arch", "a", "Arch to build the image for")
+	return c
+}
+
+// register the subcommand into rootCmd
+var _ = NewBuildISO(rootCmd, true)
```

Comment:
I might be mistaken but we expect `rootCmd` to be already set ([by this code](https://github.com/kairos-io/osbuilder/pull/70/files#diff-6913ed1d24917a2c466638c753b1382758829434826739964ce1b5051ea0b620R35)) because golang will parse `root.go` first for some reason (I'm not even sure this is happening). This is brittle and afaict all cobra examples set the rootCmd as a declaration at the top level (like we do) but add subcommands inside an `init()` function. This is probably to guarantee that rootCmd is set before we add any subcmomands to it ([relevant example](https://stackoverflow.com/a/24790378)).

Another thing is the name of this function: `NewBuildISO`.  It reads as if it would just return the command object. Adding it to the rootCmd (or any command in that regard) should be the callers responsibility. Combined with the previous comment, the caller would be an `init()` function.

Reply:
yes, but this is done this way so you can test the commands directly without having to go overcomplicated with tests IIRC. This makes it more testable that using an init function. 

Also this has been like this on elemental-cli for a couple of years, never had any issue. What I mean by this is, this has always worked, never show any problems, so Im not seeing what could go wrong here or why we should now change it and what justifies changing something that works. Can you give me a real example of a possible issue?

I think the root.go file is special and the first one to be loaded by cobra, which makes this work with no issues as long as you have your root.go.


I can change it I dont mind, I just done see the advantages of it, especially when this is (or its supposed to be) a transition binary for elemental-cli until we integrate the libs into auroraboot+osbuilder-controller directly, then the actual binary can be dropped and only the libs will live?

Reply:
See https://github.com/kairos-io/elemental-cli/blob/main/cmd/build-iso_test.go for the testing stuff!

Reply:
I'm not sure if this makes it clear and I'm not exactly sure why or when it works and when it doesn't. But at least this proves you can get funny results:

https://gist.github.com/jimmykarily/8aff809e31c3d6d0e76c4a31ebcc2f63